core/src/rpc_subscriptions.rs

```diff
@@ -132,53 +138,53 @@ where
     K: Eq + Hash + Clone + Copy,
     S: Clone + Serialize,
     B: Fn(&Bank, &K) -> X,
-    F: Fn(X, u64) -> Box<dyn Iterator<Item = S>>,
-    X: Clone + Serialize,
+    F: Fn(X, Slot) -> (Box<dyn Iterator<Item = S>>, Slot),
+    X: Clone + Serialize + Default + std::fmt::Debug,
 {
-    let mut confirmation_slots: HashMap<usize, Slot> = HashMap::new();
     let r_block_commitment_cache = block_commitment_cache.read().unwrap();
     let current_slot = r_block_commitment_cache.slot();
-    let root = r_block_commitment_cache.root();
-    let current_ancestors = bank_forks
-        .read()
-        .unwrap()
-        .get(current_slot)
-        .unwrap()
-        .ancestors
-        .clone();
-    for (slot, _) in current_ancestors.iter() {
-        if let Some(confirmations) = r_block_commitment_cache.get_confirmation_count(*slot) {
-            confirmation_slots.entry(confirmations).or_insert(*slot);
-        }
-    }
+    let node_root = r_block_commitment_cache.root();
+    let largest_confirmed_root = r_block_commitment_cache.largest_confirmed_root();
     drop(r_block_commitment_cache);
 
     let mut notified_set: HashSet<SubscriptionId> = HashSet::new();
     if let Some(hashmap) = subscriptions.get(hashmap_key) {
-        for (sub_id, (sink, confirmations)) in hashmap.iter() {
-            let desired_slot = if *confirmations == 0 {
-                Some(&current_slot)
-            } else if *confirmations == MAX_LOCKOUT_HISTORY + 1 {
-                Some(&root)
-            } else {
-                confirmation_slots.get(confirmations)
+        for (
+            sub_id,
+            SubscriptionData {
+                sink,
+                commitment,
+                last_notified_slot,
+            },
+        ) in hashmap.iter()
+        {
+            let mut slot = match commitment.commitment {
+                CommitmentLevel::Max => largest_confirmed_root,
+                CommitmentLevel::Recent => current_slot,
+                CommitmentLevel::Root => node_root,
             };
-            if let Some(&slot) = desired_slot {
-                let results = {
-                    let bank_forks = bank_forks.read().unwrap();
-                    let desired_bank = bank_forks.get(slot).unwrap();
-                    bank_method(&desired_bank, hashmap_key)
-                };
-                for result in filter_results(results, root) {
-                    notifier.notify(
-                        Response {
-                            context: RpcResponseContext { slot },
-                            value: result,
-                        },
-                        sink,
-                    );
-                    notified_set.insert(sub_id.clone());
-                }
+            let results = {
+                let bank_forks = bank_forks.read().unwrap();
+                bank_forks
+                    .get(slot)
+                    .map(|desired_bank| bank_method(&desired_bank, hashmap_key))
```

Comment:
what do you think about combining `bank_method` and `filter_results` into one function? Their responsibilities are a little fuzzy to me now. I think one function like `bank_results` would simplify the code here quite a bit

Reply:
Hmm, `bank_method` is literally the runtime Bank method, like `Bank::get_account_modified_slot`, that returns account or signature data. So it can't really be combined with `filter_results`. It could, I suppose, be called in another `bank_results` function. I'm not really fond of the way filter results works, so I'll give that a go and see if it's simpler.

Reply:
I gave this a go, but didn't commit, as it resulted in more duplication (3x https://github.com/solana-labs/solana/blob/033c697bff805b33f1607024763e775914a7fb88/core/src/rpc_subscriptions.rs#L167-L189, for each of account/program/signature). It might be less cognitive load, though.

Reply:
Ok, we can revisit later!